nemo/collections/asr/parts/preprocessing/segment.py

```diff
@@ -261,22 +262,38 @@ def segment_from_file(
 
         Note that audio_file can be either the file path, or a file-like object.
         """
+        is_segmented = False
         try:
             with sf.SoundFile(audio_file, 'r') as f:
                 sample_rate = f.samplerate
-                if 0 < n_segments < len(f):
-                    max_audio_start = len(f) - n_segments
+                if target_sr is not None:
+                    n_segments_at_original_sr = math.ceil(n_segments * sample_rate / target_sr)
+                else:
+                    n_segments_at_original_sr = n_segments
+
+                if 0 < n_segments_at_original_sr < len(f):
+                    max_audio_start = len(f) - n_segments_at_original_sr
                     audio_start = random.randint(0, max_audio_start)
                     f.seek(audio_start)
-                    samples = f.read(n_segments, dtype='float32')
+                    samples = f.read(n_segments_at_original_sr, dtype='float32')
+                    is_segmented = True
+                elif n_segments_at_original_sr >= len(f):
+                    logging.warning(
+                        f"Number of segments is greater than the length of the audio file {audio_file}. This may lead to shape mismatch errors."
+                    )
+                    samples = f.read(dtype='float32')
                 else:
                     samples = f.read(dtype='float32')
```

Comment:
Does this elif need to be here? What does the elif cover that the else doesnt?

Reply:
else covers n_segments_at_original_sr == 0. I think n_segments=0 means do not segment the audio (this was the original behaviour and was present before I made the changes). The elif is there to warn the user that the requested number of samples is more than the length of the audio file. These changes in segment.py have already been merged in the main branch in a different PR. @redoctopus and @XuesongYang reviewed that PR.